### PR TITLE
feat(iroh): allow custom protocols with optional manual accept

### PR DESCRIPTION
**Pull request by Frando**
## Description

allow to manually accept connections and use custom protocols together with iroh core

see the included example which uses both a custom protocol and core iroh protocols.

## Change checklist

- [ ] Self-review.
- [ ] Documentation updates if relevant.
- [ ] Tests if relevant.
- [ ] All breaking changes documented.

### CODE REVIEW

File: iroh/src/node/builder.rs [hunk 1]
```diff
@@ -328,6 +350,19 @@ where
...
 
+    /// Set the mode in which the node accepts incoming connections.
+    ///
+    /// If set to [`AcceptMode::Manual`], incoming connections have to be accepted by the
+    /// application code in a loop. Connections that should be handled by iroh can be passed to
+    /// [`Node::handle_connection`]. This allows to handle custom protocols.
+    ///
+    /// The default is [`AcceptMode::Auto`], which accepts all iroh protocol automatically and does
+    /// not allow to use custom protocols.
+    pub fn accept_mode(mut self, accept_mode: AcceptMode) -> Self {
```
Comment (rklaehn) [inline iroh/src/node/builder.rs:361]: I like passing the mode instead of having accept_manual().

**Review by rklaehn — COMMENTED** [review]
I like this approach better than some elaborate scheme with ALPN handlers. I think the complexity is acceptable for advanced use cases, and having some handler trait scheme will break down in many cases.

One thing I was wondering: you can already get the endpoint from the node. So what would happen if somebody would just grab an endpoint and then call accept on it, in the current code? Who wins? Is it round robin?

https://docs.rs/iroh/latest/iroh/node/struct.Node.html#method.magic_endpoint

**Comment by Frando** [comment]
> One thing I was wondering: you can already get the endpoint from the node. So what would happen if somebody would just grab an endpoint and then call accept on it, in the current code? Who wins? Is it round robin?

Yes, it would be round-robin I think, same as `quinn::Endpoint::accept`. 

You could already do this now by the way - the endpoint is already exposed (as you linked). You really shouldn't though because behavior would be unpredictable.

We could remove Node::accept in the PR, it would be identical to calling node.magic_endpoint().accept.

**Comment by Frando** [comment]
Closing in favor of #2357

**Pull request closed without merging**